cmd/registry/breaking-change-detector/breaking_change_detector.go

```diff
@@ -0,0 +1,212 @@
+			Modifications: make(map[string]*rpc.Diff_ValueChange),
+		},
+	}
+	allChanges.BreakingChanges = getBreakingChanges(diff)
+	allChanges.NonBreakingChanges = getNonBreakingChanges(diff)
+	allChanges.UnknownChanges = getUnknownChanges(diff)
+	return allChanges
+}
+
+func getBreakingChanges(diff *rpc.Diff) *rpc.Diff {
+	breakingChanges := &rpc.Diff{
+		Additions:     []string{},
+		Deletions:     []string{},
+		Modifications: map[string]*rpc.Diff_ValueChange{},
+	}
+	for _, addition := range diff.GetAdditions() {
+		if fitsAnyPattern(unsafeAdds, addition) {
+			breakingChanges.Additions = append(breakingChanges.Additions, addition)
+		}
+	}
+
+	for _, deletion := range diff.GetDeletions() {
+		if fitsAnyPattern(unsafeDeletes, deletion) {
+			breakingChanges.Deletions = append(breakingChanges.Deletions, deletion)
+		}
+	}
+
+	for modification, modValue := range diff.GetModifications() {
+		if fitsAnyPattern(unsafeMods, modification) {
+			breakingChanges.Modifications[modification] = modValue
+		}
+	}
+	return breakingChanges
+}
+
+func getNonBreakingChanges(diff *rpc.Diff) *rpc.Diff {
+	nonBreakingChanges := &rpc.Diff{
+		Additions:     []string{},
+		Deletions:     []string{},
+		Modifications: map[string]*rpc.Diff_ValueChange{},
+	}
+	for _, addition := range diff.GetAdditions() {
+		if fitsAnyPattern(safeAdds, addition) {
+			nonBreakingChanges.Additions = append(nonBreakingChanges.Additions, addition)
+		}
+	}
+	for _, deletion := range diff.GetDeletions() {
+		if fitsAnyPattern(safeDeletes, deletion) {
+			nonBreakingChanges.Deletions = append(nonBreakingChanges.Deletions, deletion)
+		}
+	}
+
+	for modification, modValue := range diff.GetModifications() {
+		if fitsAnyPattern(safeMods, modification) {
+			nonBreakingChanges.Modifications[modification] = modValue
+		}
+	}
+	return nonBreakingChanges
+}
+
+func getUnknownChanges(diff *rpc.Diff) *rpc.Diff {
+	unknownChanges := &rpc.Diff{
+		Additions:     []string{},
+		Deletions:     []string{},
+		Modifications: map[string]*rpc.Diff_ValueChange{},
+	}
+	for _, addition := range diff.GetAdditions() {
+		if !fitsAnyPattern(safeAdds, addition) && !fitsAnyPattern(unsafeAdds, addition) {
+			unknownChanges.Additions = append(unknownChanges.Additions, addition)
+		}
+	}
+
+	for _, deletion := range diff.GetDeletions() {
+		if !fitsAnyPattern(safeDeletes, deletion) && !fitsAnyPattern(unsafeDeletes, deletion) {
+			unknownChanges.Deletions = append(unknownChanges.Deletions, deletion)
+		}
+	}
+
+	for modification, modValue := range diff.GetModifications() {
+		if !fitsAnyPattern(safeMods, modification) && !fitsAnyPattern(unsafeMods, modification) {
+			unknownChanges.Modifications[modification] = modValue
+		}
+	}
+	return unknownChanges
+}
+
+func fitsAnyPattern(patterns []detectionPattern, change string) bool {
+	for _, pattern := range patterns {
+		if pattern.PositiveMatchRegex == nil {
+			if !pattern.NegativeMatchRegex.MatchString(change) {
+				return true
+			}
+			continue
+		}
```

Comment:
This case doesn't seem to exist with the current set of patterns, since a positive pattern is always provided. Is it safe to assume the negative match regex is non-nil here, or what happens in that case?

How do these checks differ from the following code?
```
		if p := pattern.NegativeMatchRegex; p != nil && p.MatchString(change) {
			continue
		}
		if p := pattern.PositiveMatchRegex; p != nil && p.MatchString(change) {
			return true
		}
```

Reply:
Done